File: Codes/gracekoo/21_merge-two-sorted-lists.py
# -*- coding: utf-8 -*-
# @Time: 2020/1/1 9:04 下午
# @Author: GraceKoo
# @File: 21_merge-two-sorted-lists.py
# @Desc: https://leetcode-cn.com/problems/merge-two-sorted-lists/


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head_pre = ListNode(-1)
        pre = head_pre
        while l1 or l2:
            if l1.value < l2.value:
                pre.next = l1
                l1 = l1.next
            else:
                pre.next = l2
                l2 = l2.next
            # pre指向当前计算节点的上一个节点，运算结束后需指向下一个
            pre = pre.next
        pre.next = l1 if l1 is not None else l2

        return head_pre.next

File: Codes/yore/21_merge_two_sorted_lists/21_merge_two_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode root = new ListNode(-1);
        ListNode point = root;
        int data1;
        int data2;
        while (l1 != null || l2 != null) {
            if (l1 != null && l2 != null) {
                data1 = l1.val;
                data2 = l2.val;
                if (data1 < data2) {
                    point.next = new ListNode(data1);
                    l1 = l1.next;
                } else {
                    point.next = new ListNode(data2);
                    l2 = l2.next;
                }
                point.next = new ListNode(data1 < data2 ? data1 : data2);

            } else if (l1 != null) {
                point.next = new ListNode(l1.val);
                l1 = l1.next;
            } else {
                point.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            point = point.next;

        }
        return root.next;
    }
}
